### DISCUSSION TITLE
Setting state in App/index.js causes app to break on certain views when refreshing

### DISCUSSION THREAD

**Discussion by lindseysatterfield** [nss-evening-cohort-14/evening-client-side#302] [Capstones!] [answered]
#### Searched other tickets?
Yes, #252 mentioned that you cannot refresh the app on the view if you're setting state in `App/index.js`. It did not help me figure out a solution though.

#### Describe your issue
I noticed after checking out my deployed link that if I refresh on a view, the whole app breaks. Dr. T mentioned that it was because I was setting the state for `restaurants` and `setRestaurants` in `App/index.js` and prop drilling them down. Also possibly needing a useEffect in different views. I understand the reasoning why it breaks now and why we should be setting state in the view and not `App/index.js`, but unsure how to fix it. She mentioned setting state in the view that needs it, but I have three views using `setRestaurants`. `setRestaurants` is being used in `AddRestaurant.js`, `Favorites.js` and `AllRestaurants.js` to be able to perform parts of CRUD on those views. I think I just need to sit and talk this through with someone because I don't understand how to make it work in each view if I need `setRestaurants` for three separate views.

#### List any error messages that appear
Depends on what I tried but here are a few that I did get at certain points:
`Uncaught TypeError: Cannot read property 'uid' of null`
and
`index.js:1 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.`.

#### What have you googled?
Could not really find anything specific to what I'm trying to do but I did take a look at these:
https://stackoverflow.com/questions/58038008/how-to-stop-memory-leak-in-useeffect-hook-react
https://reactjs.org/docs/hooks-state.html

#### What have you tried? Tell us everything you have tried.
I tried moving state and useEffect to the views that needed it. Tried with `AllRestaurants.js` first and the app broke. Tried on the other views as well and what I tried just didn't work.

#### Link to code
https://github.com/lindseysatterfield/nashville-night-out/blob/main/src/App/index.js
https://github.com/lindseysatterfield/nashville-night-out/blob/main/src/helpers/views/AllRestaurants.js
https://github.com/lindseysatterfield/nashville-night-out/blob/main/src/helpers/views/AddRestaurant.js
https://github.com/lindseysatterfield/nashville-night-out/blob/main/src/helpers/views/Favorites.js

**Comment by TrinityChristiana** [comment]
Try recreating your flow chart (or creating one if you don't already have one) with your new state and prop flow. List out each view and component you have and show where  the following thing will happen:

1. Where each state is being instantiated (Where the `useState` hook is being used)
2. Where each state item (function and value) is being passed down as props
3. Where the set state function is being called and where the state value is being used.

It will be a great way to start over without having to rewrite your code. Then go view by view updating state and your props.

**Comment by lindseysatterfield** [comment]
On my local machine, when I refresh on a view, it takes me back to the home view. So I cannot tell if my app is breaking or not on that view since it sends me back. When I refresh on a view that is not the home view on my deployed version, I get a "Page Not Found." So while the app is working, refreshing on a view other than the home page is giving me a "Page Not Found."

Working on trying to move useEffects and state into views now and not in `App/index.js`.

**Comment by drteresavasquez** [comment]
@lindseysatterfield is working through this

**Comment by lindseysatterfield** [comment, marked as answer]
Update: currently still working through state issues but did fix the Netlify "Page Not Found" error thanks to this ticket #248. Completely forgot that I needed that file since I'm using private routes. Now I no longer get that error and I can see what my app looks like. Cards are not there yet when refreshing on deployed views, but I think that is due to still being in the middle of my state issue.

**Comment by lindseysatterfield** [comment]
Another update: I have all state set out of App/index.js and only in views now. Mostly everything is working! But I cannot get the DOM to rerender on the Favorites view after making an update or delete on Favorites view. Must refresh to see changes. All of this was working previously before when I had state set up in App/index.js and now after moving things, it no longer works.

I think the problem is happening here on line 60:
https://github.com/lindseysatterfield/nashville-night-out/blob/state-fix/src/components/RestaurantForm.js#L60

**Reply by drteresavasquez** [reply]
@hmparsons-sde, @danicrosby  you are having something similar I believe and have it working in some capacity in your app. Can you put your heads together to help each other out here?

**Reply by drteresavasquez** [reply]
Also, this is likely a new ticket. I feel like we resolved the initial issue on this ticket before.

**Reply by lindseysatterfield** [reply]
@drteresavasquez After working through tonight with Aja, it seems to be an issue with my `if` statement and the logic inside. Will be working through that and will make another ticket if needed.

**Reply by drteresavasquez** [reply]
I see. Is the second code block ever running? It would seem the first code block would be true on any update?

**Reply by drteresavasquez** [reply]
The message above is still true so check that second code block as well after trying the suggested solution below. 

Wait. I think I know why this is happening. You're passing setRestaurants in The favorites view but you're never using the value because you don't even name it in the destructure. You should instead be passing setFavorites. To make this work for your component, you can change 

setRestaurants={setRestaurants}

TO:

setRestaurants={setFavorites}

On this line: https://github.com/lindseysatterfield/nashville-night-out/blob/257b3c4483b55719cb5c1ffa3dc0e1e376f6de67/src/helpers/views/Favorites.js#L50

Do not change anything else other than that line and see if it works. 

YES I know the prop you're passing is named setRestaurants but that's what you're using in your form component as a prop. The value needs to be setFavorites in order for it to work. 

If this doesn't work, let's create a new ticket please.

**Reply by hmparsons-sde** [reply]
So sorry, just seeing this. I was not passing my equivalent of "setFavorites" everywhere that I needed to (what @drteresavasquez was saying). I also had not added it to my "handleClick/handleSubmit" functions. Mine is called "setTripLocations":

handleClick for "Delete":
`deleteLocation(tripLocation.firebaseKey, tripLocation.uid)
.then(() => getTripLocation(tripLocation.tripId))
.then((response) => setTripLocations(response));`

handleAdd for "Create":
`const handleAdd = (e) => {    
e.preventDefault();    
const locationObj = {      
cityName: pressureObj.name,      
pressure: pressureObj.main.pressure,      
tripId: firebaseKey,      
uid    
};    
console.warn(locationObj);   
 addLocation(locationObj, uid)      
.then(() => getTripLocation(locationObj.tripId))      
.then((response) => {        
console.warn(response);       
 setTripLocations(response);      
});  };`

**Reply by lindseysatterfield** [reply]
@drteresavasquez Tried your fix and that just gave me an error of `Line 9:12:  'setRestaurants' is assigned a value but never used`

You asked above if the second block of code ever runs. On my deployed app, everything updates/deletes, and works perfect in all views, including `Favorites.js` with that code. It was when I started moving state down from `App/index.js` that seems to have broken `Favorites.js` update/delete on my branch. Will definitely be making another ticket if I cannot figure it out!

**Reply by drteresavasquez** [reply]
Oh. You just needed to remove the state item for setRestaurants. By making the change it was no longer necessary.
